// @histoire/app/dist/bundled/components/app/Breadcrumb.js
import { pushScopeId, popScopeId, createElementVNode, createTextVNode, defineComponent, useCssVars, unref, computed, ref, watch, openBlock, createElementBlock, Fragment, renderList, toDisplayString, createVNode, normalizeClass, withCtx } from "@histoire/vendors/vue";
import { Icon } from "@histoire/vendors/iconify";
import { useStoryStore } from "../../stores/story.js";
import _sfc_main$1 from "./MobileOverlay.js";
import _sfc_main$2 from "../tree/StoryList.js";
import "./Breadcrumb.vue_vue_type_style_index_0_scoped_true_lang.js";
import _export_sfc from "../../_virtual/plugin-vue_export-helper.js";
const _withScopeId = (n) => (pushScopeId("data-v-25f7510c"), n = n(), popScopeId(), n);
const _hoisted_1 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createElementVNode("span", { class: "htw-opacity-40" }, " / ", -1));
const _hoisted_2 = { class: "htw-flex htw-items-center htw-gap-2" };
const _hoisted_3 = { class: "htw-opacity-40 htw-text-sm" };
const _hoisted_4 = /* @__PURE__ */ createTextVNode(" Select a story... ");
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "Breadcrumb",
  props: {
    tree: null,
    stories: null
  },
  setup(__props) {
    useCssVars((_ctx) => {
      var _a;
      return {
        "353b213f": (_a = unref(story)) == null ? void 0 : _a.iconColor
      };
    });
    const storyStore = useStoryStore();
    const story = computed(() => storyStore.currentStory);
    const folders = computed(() => {
      return story.value.file.path.slice(0, -1);
    });
    const isMenuOpened = ref(false);
    function openMenu() {
      isMenuOpened.value = true;
    }
    function closeMenu() {
      isMenuOpened.value = false;
    }
    watch(story, () => {
      isMenuOpened.value = false;
    });
    return (_ctx, _cache) => {
      var _a;
      return openBlock(), createElementBlock(Fragment, null, [
        createElementVNode("div", null, [
          createElementVNode("a", {
            class: "htw-px-6 htw-h-12 hover:htw-text-primary-500 dark:hover:htw-text-primary-400 htw-cursor-pointer htw-flex htw-gap-2 htw-flex-wrap htw-w-full htw-items-center",
            onClick: openMenu
          }, [
            unref(story) ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(unref(folders), (file, key) => {
                return openBlock(), createElementBlock(Fragment, { key }, [
                  createElementVNode("span", null, toDisplayString(file), 1),
                  _hoisted_1
                ], 64);
              }), 128)),
              createElementVNode("span", _hoisted_2, [
                createVNode(unref(Icon), {
                  icon: (_a = unref(story).icon) != null ? _a : "carbon:cube",
                  class: normalizeClass(["base-list-item-link-icon htw-w-5 htw-h-5 htw-flex-none", {
                    "htw-text-primary-500": !unref(story).iconColor,
                    "bind-icon-color": unref(story).iconColor
                  }])
                }, null, 8, ["icon", "class"]),
                createTextVNode(" " + toDisplayString(unref(story).title) + " ", 1),
                createElementVNode("span", _hoisted_3, toDisplayString(unref(story).variants.length), 1)
              ])
            ], 64)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
              _hoisted_4
            ], 64)),
            createVNode(unref(Icon), {
              icon: "carbon:chevron-sort",
              class: "htw-w-5 htw-h-5 htw-shrink-0 htw-ml-auto"
            })
          ])
        ]),
        createVNode(_sfc_main$1, {
          title: "Select a story",
          opened: isMenuOpened.value,
          onClose: closeMenu
        }, {
          default: withCtx(() => [
            createVNode(_sfc_main$2, {
              tree: __props.tree,
              stories: __props.stories,
              class: "htw-flex-1 htw-overflow-y-scroll"
            }, null, 8, ["tree", "stories"])
          ]),
          _: 1
        }, 8, ["opened"])
      ], 64);
    };
  }
});
var Breadcrumb = /* @__PURE__ */ _export_sfc(_sfc_main, [["__scopeId", "data-v-25f7510c"]]);
export { Breadcrumb as default };
