// @histoire/app/dist/bundled/components/toolbar/StoryVariantBackground.js
import { pushScopeId, popScopeId, createElementVNode, createTextVNode, defineComponent, useCssVars, unref, resolveComponent, resolveDirective, openBlock, createBlock, withCtx, createVNode, createElementBlock, Fragment, renderList, normalizeClass, toDisplayString, normalizeStyle, createCommentVNode, withDirectives } from "@histoire/vendors/vue";
import { Icon } from "@histoire/vendors/iconify";
import { usePreviewSettingsStore } from "../../stores/preview-settings.js";
import { histoireConfig } from "../../util/config.js";
import _sfc_main$1 from "../base/BaseCheckbox.js";
import "./StoryVariantBackground.vue_vue_type_style_index_0_scoped_true_lang.js";
import _export_sfc from "../../_virtual/plugin-vue_export-helper.js";
const _withScopeId = (n) => (pushScopeId("data-v-bc4883a4"), n = n(), popScopeId(), n);
const _hoisted_1 = { class: "htw-cursor-pointer hover:htw-text-primary-500 htw-flex htw-items-center htw-gap-1 htw-h-full htw-px-2 htw-group" };
const _hoisted_2 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createElementVNode("div", { class: "bind-preview-bg htw-w-4 htw-h-4 htw-rounded-full htw-border htw-border-black/50 dark:htw-border-white/50" }, null, -1));
const _hoisted_3 = { class: "htw-flex htw-flex-col htw-items-stretch" };
const _hoisted_4 = /* @__PURE__ */ createTextVNode(" Checkerboard ");
const _hoisted_5 = ["onClick"];
const _hoisted_6 = { class: "htw-mr-auto" };
const _hoisted_7 = { class: "htw-ml-auto htw-opacity-70" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "StoryVariantBackground",
  setup(__props) {
    useCssVars((_ctx) => ({
      "6bcdf233": unref(settings).backgroundColor
    }));
    const settings = usePreviewSettingsStore().currentSettings;
    return (_ctx, _cache) => {
      const _component_VDropdown = resolveComponent("VDropdown");
      const _directive_tooltip = resolveDirective("tooltip");
      return unref(histoireConfig).backgroundPresets.length ? (openBlock(), createBlock(_component_VDropdown, {
        key: 0,
        placement: "bottom-end",
        skidding: 6,
        class: "htw-h-full htw-flex-none"
      }, {
        popper: withCtx(({ hide }) => [
          createElementVNode("div", _hoisted_3, [
            createVNode(_sfc_main$1, {
              modelValue: unref(settings).checkerboard,
              "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => unref(settings).checkerboard = $event)
            }, {
              default: withCtx(() => [
                _hoisted_4
              ]),
              _: 1
            }, 8, ["modelValue"]),
            (openBlock(true), createElementBlock(Fragment, null, renderList(unref(histoireConfig).backgroundPresets, (option, index) => {
              return openBlock(), createElementBlock("button", {
                key: index,
                class: normalizeClass(["htw-px-4 htw-py-3 htw-cursor-pointer htw-text-left htw-flex htw-gap-4", [
                  unref(settings).backgroundColor === option.color ? "htw-bg-primary-500 hover:htw-bg-primary-600 htw-text-white dark:htw-text-black" : "htw-bg-transparent hover:htw-bg-primary-100 dark:hover:htw-bg-primary-700"
                ]]),
                onClick: ($event) => {
                  unref(settings).backgroundColor = option.color;
                  hide();
                }
              }, [
                createElementVNode("span", _hoisted_6, toDisplayString(option.label), 1),
                option.color !== "$checkerboard" ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
                  createElementVNode("span", _hoisted_7, toDisplayString(option.color), 1),
                  createElementVNode("div", {
                    class: "htw-w-4 htw-h-4 htw-rounded-full htw-border htw-border-black/20 dark:htw-border-white/20",
                    style: normalizeStyle({
                      backgroundColor: option.color
                    })
                  }, null, 4)
                ], 64)) : createCommentVNode("", true)
              ], 10, _hoisted_5);
            }), 128))
          ])
        ]),
        default: withCtx(() => [
          withDirectives((openBlock(), createElementBlock("div", _hoisted_1, [
            _hoisted_2,
            createVNode(unref(Icon), {
              icon: "carbon:caret-down",
              class: "htw-w-4 htw-h-4 htw-opacity-50 group-hover:htw-opacity-100"
            })
          ])), [
            [_directive_tooltip, "Background color"]
          ])
        ]),
        _: 1
      })) : createCommentVNode("", true);
    };
  }
});
var StoryVariantBackground = /* @__PURE__ */ _export_sfc(_sfc_main, [["__scopeId", "data-v-bc4883a4"]]);
export { StoryVariantBackground as default };
